packages/navigator/src/browser/navigator-widget.tsx

```diff
@@ -213,6 +224,23 @@ export class FileNavigatorWidget extends FileTreeWidget {
         </div>;
     }
 
+    /**
+     * When a multi-root workspace is opened, a user can remove all the folders from it.
```

Comment:
Sorry, I don't get the nuance? In the scenario described, despite having no roots at all, we are still in a state where a multi-root workspace is active (from the `WorkspaceService` perspective).

Reply:
TLDR: 'multi-root' is fine, but not exact, since the distinction isn't about the number of roots but the presence or absence of a workspace file.

You're right that the WorkspaceService doesn't distinguish clearly, but I think that's a problem on the part of the workspace service. There are two distinctions: single-root vs. multi-root, and saved vs. unsaved. All multi-root workspaces are saved workspaces in that they're defined in a workspace file (even if that file hasn't been saved to disk), but a workspace with one root or no roots can also be defined in a workspace file. The workspace service is a little bit fuzzy on the distinction, and I believe `isMultiRootWorkspaceOpened` returns true for all saved workspaces, even if there is only one root. That isn't ideal, because there are times when you might prefer to let the actual number of workspace roots determine behavior, but if you rely on the `isMultiRootWorkspaceOpened` boolean rather than checking the actual number of roots, you might not select the intended behavior. In this case, you're interested in cases where a saved workspace is open, but contains no roots.

Reply:
Could you expand on the "saved vs. unsaved" distinction?

My understanding right now is that we only have a distinction between single-root vs multi-root, and my approach was to check the tree model to find the difference. I think I got the logic wrong though, so I'm looking into what I might have missed.

Reply:
It really isn't that important for the current context, and I don't think it has any bearing on your code - it's just a leftover peccadillo from thinking about scopes for the preference UI work. Basically, if you open a folder as a workspace then it is an unsaved workspace. The workspace service has a getter that returns that fact:

https://github.com/eclipse-theia/theia/blob/5eb6efa40129101b74a88a16e0c45e81545ab966/packages/workspace/src/browser/workspace-service.ts#L523-L525

the check for `isDirectory` there will return false if the workspace is stored in a workspace folder, because a workspace stored in a file counts as a 'saved' workspace, no matter how many roots it has. It turns out that the getter for `isMultiRootWorkspaceOpened` is actually exactly the same:

https://github.com/eclipse-theia/theia/blob/5eb6efa40129101b74a88a16e0c45e81545ab966/packages/workspace/src/browser/workspace-service.ts#L307-L313

But that means that `multi-root` is a misnomer, because it doesn't actually mean that your workspace has more than one root at all. Probably the code for for `isMultiRootWorkspaceOpened` should be something like

```typescript
    get isMultiRootWorkspaceOpened(): boolean {
        return !!this.workspace && !this.workspace.isDirectory && this.tryGetRoots().length > 1; // or !== 1, depending
    }
```

Reply:
Yes, I think it's too late to get the terminology straight now. As @marechal-p jokes, it should really be `isPotentiallyMultiRootWorkspaceOpened` 😄.